### DISCUSSION TITLE
supabase auth provide with one tap google login

### DISCUSSION THREAD

**Discussion by M-NaveedAshraf-work** [supabase/supabase#8035] [Questions] [answered]
I'm actually trying to auth my users with supabase with google on tap login provider. it's working fine for a user having one google account register in a browser. but if a user has multiple accounts in a browser then the supabase auth function redirects to google  select an account page. However, the user already selected an account in one tap login iframe.

here I selected a user to login
![InkedWMh18B452VZh](https://user-images.githubusercontent.com/107533241/182521155-8422fd32-8741-4514-8879-54ad24b44d02.jpg)
 
and it redirects me to this page.
![Inked0jja68vxyWQN](https://user-images.githubusercontent.com/107533241/182521393-b98c8104-7470-471a-9b47-c0343b7c48ee.jpg)

**Comment by projectashik** [comment]
Can you share the source code?

**Reply by M-NaveedAshraf-work** [reply]
```
async function loginWithGoogle() {

     const { user, session, error } = await supabase.auth.signIn({
              // provider can be 'github', 'google', 'gitlab', or 'bitbucket'
              provider: 'google'
     })
}
```

This `ShowOneTapGoogleLogin` function called on page load and the I call a callback function (`loginWithGoogle`) when user click on an account to coninue login.

```
const ShowOneTapGoogleLogin = () => {
    setTimeout(() => {
      window.google.accounts.id.initialize({
        client_id: 'my_client_id',
        callback: loginWithGoogle,
      });
      window.google.accounts.id.renderButton(
        document.getElementById("buttonDiv"),
        { theme: "outline", size: "large" }  // customization attributes
      );
      window.google.accounts.id.prompt((notification) => {
        console.log('this is the prompt notification', notification);
      });
    }, 5000);
  }
```

**Comment by M-NaveedAshraf-work** [comment]
As I can see there's a new release v 2.0 by supabase. Is there any update about this issue?

**Comment by KULTI1995** [comment]
really, nothing has moved on this topic? just enter login, using the token, specifying provider. Please bring this in at last.

**Comment by pixelkoduje** [comment]
Excuse me, but will we ever get an answer? Implementing it takes a few days of work, and you can clearly see the demand.

Do we have to be investors and give big money to get any answer? Because it's not serious.

**Reply by moarsel** [reply]
Accusing the developers of a conspiracy because they didn't implement your favourite feature is unprofessional (and comes across as rude). This issue only has 5 upvotes, and I trust they're focusing on delivering value to as many ppl as possible. If you're stuck you can ask questions here or on stackOverflow, or dig into the codebase yourself to see what's going on -- you'll learn a lot that way!

**Reply by madebyisaacr** [reply]
Supabase is free and open source, they aren't obligated to reply or to add a feature you want. If you need this, you can edit the code and submit a pull request yourself or hire someone else to do it.

**Reply by KULTI1995** [reply]
This is a functionality that exists on tens of millions of sites, which helps get more users, just no sites put on supabase. That's why I'm giving up on it, even though it was initially one of my favorite pieces of software.

**Comment by moarsel** [comment]
Thanks for the starter code @M-NaveedAshraf-work! Your approach works because we're effectively running the oAuth process twice, first with Google's sign in API, then again with supabase's. The second time through it succeeds without a dialog because the user is already signed in. But if the user has multiple google profiles, account selection can't be skipped. 

So the trick is to use the credentials returned in the first auth, get the user's email and send it along for the second pass. 

```
window.google.accounts.id.initialize({
      client_id: GOOGLE_CLIENT_ID,
      callback: loginWithGoogle,
    });
    
  // credential jwt is passed to the callback when user logs in
 async function loginWithGoogle({ credential }) {
    // use `jwt-decode` package to get the email address from the credential jwt.
    const decoded = jwtDecode(credential);
    await supabase.auth.signInWithOAuth({
      provider: "google",
      options: {
        // google oAuth takes a 'login_hint' query param if we already know which email we're logging into
        queryParams: { login_hint: decoded.email },
      },
    });
  }
```

It's too bad there isn't an out of the box function for one-tap login, or a method to pass an access_token like @KULTI1995 mentioned, but it's a quick workaround.

**Reply by TheAndrewJackson** [reply]
Thanks for this snippet! It worked for me. 

I'm running into one small issue that you may have seen. This snippet causes a full page refresh in order to get the session due to redirecting to google and back. I experimented with removing the redirect but it looks like supabase may only support the callback redirect method. That's what their [docs ](https://supabase.com/docs/guides/auth#redirect-urls-and-wildcards)leads me to believe. I think the workaround that you posted is our best bet until supabase natively supports OAuth without redirects like google one tap.

Just in case... do you know of a way to get the full supabase user session through oauth without the full page refresh/ redirect flow?  

------

### Experiment results
I think this is happening because it still redirects to googles auth page/route where the user would normally select there email and then gets sent back. The login hint helps but it doesn't remove that step. 

I tried setting the `skipBrowserRedirect` option to true but that prevents it from redirecting to google and getting the full session.
```typescript
const OauthResponse = await supabase.auth.signInWithOAuth({
      provider: "google",
      options: {
        skipBrowserRedirect: true,
        // google oAuth takes a 'login_hint' query param if we already know which email we're logging into
        queryParams: { login_hint: decoded.email },
      },
    });
```

`signInWithOAuth` returns an object that looks like this
```json
{
  "data": {
      "provider": "google"
      "url": "your callback url"
    }
  "error": null
}
```

The url is what's used to start the auth redirect chain. If called from postman it will return the google auth page where the user selects their email. If called from a new tab in the browser it will redirect back to your app.

This leads me to believe that supabase only supports OAuth by using the redirect method.

**Comment by LehuyH** [comment, marked as answer]
One tap is now supported
https://supabase.com/blog/native-mobile-auth#sign-in-with-apple-js-google-one-tap-and-chrome-extensions

**Reply by davidkhierl** [reply]
Hi! I have questions about implementing this with nextjs?

this part here is a bit confusing. how can I register this callback method?

![image](https://github.com/supabase/supabase/assets/25083033/f9a6d340-6c2b-4669-980e-eb9c6d7f3156)

**Reply by matiasbattocchia** [reply]
This worked for me!

```ts
  async function handleSignInWithGoogle(response: any) {
    const { data, error } = await supabase.auth.signInWithIdToken({
      provider: 'google',
      token: response.credential,
    })
  }

  useEffect(() => {
    const script = document.createElement("script")
    script.src = "https://accounts.google.com/gsi/client"
    script.async = true
    script.defer = true
    script.onload = () => {
      // @ts-expect-error
      window.google.accounts.id.initialize({
        client_id: '<your_client_id>',
        callback: handleSignInWithGoogle
      });
      // @ts-expect-error
      window.google.accounts.id.prompt()
    }

    document.body.appendChild(script)
  }, [])
```

Source: https://es.slideshare.net/SufalamTechnologies/how-to-implement-google-one-tap-login-in-reactjs
